src/run/regional-blob-store.cts

```diff
@@ -53,10 +68,131 @@ const fetchBeforeNextPatchedItFallback = forceOptOutOfUsingDataCache(
 const getFetchBeforeNextPatchedIt = () =>
   extendedGlobalThis[FETCH_BEFORE_NEXT_PATCHED_IT] ?? fetchBeforeNextPatchedItFallback
 
-export const getRegionalBlobStore = (args: GetWithMetadataOptions = {}): Store => {
+const getRegionalBlobStore = (args: GetWithMetadataOptions = {}): Store => {
   return getDeployStore({
     ...args,
     fetch: getFetchBeforeNextPatchedIt(),
     region: process.env.USE_REGIONAL_BLOBS?.toUpperCase() === 'TRUE' ? undefined : 'us-east-2',
   })
 }
+const DEFAULT_FALLBACK_MAX_SIZE = 50 * 1024 * 1024 // 50MB, same as default Next.js config
+export function setInMemoryCacheMaxSizeFromNextConfig(size: unknown) {
+  if (typeof size === 'number') {
+    extendedGlobalThis[IN_MEMORY_CACHE_MAX_SIZE] = size
+  }
+}
+
+function getInMemoryLRUCache() {
+  if (typeof extendedGlobalThis[IN_MEMORY_LRU_CACHE] === 'undefined') {
+    const maxSize =
+      typeof extendedGlobalThis[IN_MEMORY_CACHE_MAX_SIZE] === 'number'
+        ? extendedGlobalThis[IN_MEMORY_CACHE_MAX_SIZE]
+        : DEFAULT_FALLBACK_MAX_SIZE
+
+    extendedGlobalThis[IN_MEMORY_LRU_CACHE] =
+      maxSize === 0
+        ? null // if user sets 0 in their config, we should honor that and not use in-memory cache
+        : new LRUCache<string, BlobType | typeof NullValue | Promise<BlobType | null>>({
+            max: 1000,
+            maxSize,
+            sizeCalculation: (valueToStore) => {
+              return estimateBlobSize(valueToStore === NullValue ? null : valueToStore)
+            },
+          })
+  }
+  return extendedGlobalThis[IN_MEMORY_LRU_CACHE]
+}
+
+interface RequestSpecificInMemoryCache {
+  get(key: string): BlobType | null | Promise<BlobType | null> | undefined
+  set(key: string, value: BlobType | null | Promise<BlobType | null>): void
+}
+
+const noOpInMemoryCache: RequestSpecificInMemoryCache = {
+  get(): undefined {
+    // no-op
+  },
+  set() {
+    // no-op
+  },
+}
+
+const getRequestSpecificInMemoryCache = (): RequestSpecificInMemoryCache => {
+  const requestContext = getRequestContext()
+  if (!requestContext) {
+    // Fallback to a no-op store if we can't find request context
+    return noOpInMemoryCache
+  }
+
+  const inMemoryLRUCache = getInMemoryLRUCache()
+  if (inMemoryLRUCache === null) {
+    return noOpInMemoryCache
+  }
+
+  return {
+    get(key) {
+      const inMemoryValue = inMemoryLRUCache.get(`${requestContext.requestID}:${key}`)
+      if (inMemoryValue === NullValue) {
+        return null
+      }
+      return inMemoryValue
+    },
+    set(key, value) {
+      inMemoryLRUCache.set(`${requestContext.requestID}:${key}`, value ?? NullValue)
+    },
+  }
+}
+
+const encodeBlobKey = async (key: string) => {
+  const { encodeBlobKey: encodeBlobKeyImpl } = await import('../shared/blobkey.js')
+  return await encodeBlobKeyImpl(key)
+}
+
+export const getMemoizedKeyValueStoreBackedByRegionalBlobStore = (
```

Comment:
This looks like a whole other feature, potentially should be in another PR?

Reply:
In general I will always lean towards single-focus PRs, but let's leave things as they are this time around as your explanation gives enough context for me to review here.